add a prefix index to filecache.path, attempt 2

icewind1991 opened the pull request:
Second attempt at #26070, this time with a shorter prefix to hopefully not have the same issues.

----

The reason that `filecache.path` hasn't had an index added is the mysql limitation of ~1kb for indexeded fields,
which is to small for the `path`, however mysql supports indexing only the first N bytes of a column instead of the entire column,
allowing us to add an index even if the column is to long.

Because the index doesn't cover the entire column it can't be used in all situations where a normal index would be used, but it does cover the `path like 'folder/path/%'` queries that are used in various places.

Sqlite and Postgresql don't support prefix indexes, but they also don't have the 1kb limit and DBAL handles the differences in index creation.

Signed-off-by: Robin Appelman <robin@icewind.nl>

icewind1991 committed:
add a prefix index to filecache.path

The reason that `filecache.path` hasn't had an index added is the mysql limitation of ~1kb for indexeded fields,
which is to small for the `path`, however mysql supports indexing only the first N bytes of a column instead of the entire column,
allowing us to add an index even if the column is to long.

Because the index doesn't cover the entire column it can't be used in all situations where a normal index would be used, but it does cover the `path like 'folder/path/%'` queries that are used in various places.

Sqlite and Postgresql don't support prefix indexes, but they also don't have the 1kb limit and DBAL handles the differences in index creation.

Signed-off-by: Robin Appelman <robin@icewind.nl>

nickvergessen commented on core/Migrations/Version13000Date20170718121200.php:
```diff
@@ -263,6 +263,7 @@ public function changeSchema(IOutput $output, \Closure $schemaClosure, array $op
 			$table->addIndex(['storage', 'size', 'fileid'], 'fs_storage_size');
 			$table->addIndex(['mtime'], 'fs_mtime');
 			$table->addIndex(['size'], 'fs_size');
+			$table->addIndex(['storage', 'path'], 'fs_storage_path_prefix', [], ["lengths" => [null, 64]]);
```
```suggestion
			$table->addIndex(['storage', 'path'], 'fs_storage_path_prefix', [], ['lengths' => [null, 64]]);
```

nickvergessen commented:
Can't remember completely why it failed back then. Maybe we can ask people to setup the index manually on several instances to not fail again.

skjnldsv reviewed (approved):
Code is good, no idea about what @nickvergessen was referring to. Let's wait for another review :)

icewind1991 committed:
disable path prefix index on postgresql for now

having the index work properly for the queries we need it for requires some additional options which dbal does not support at the momement.
to prevent making it harder to add the correct index later on we don't create the index for now on postgresql

Signed-off-by: Robin Appelman <robin@icewind.nl>

skjnldsv commented:
Any news? :)

nickvergessen commented:
Seems it's helping lot of times, but e.g. with groupfolders the index is not used:

EXPLAIN SELECT `file`.`fileid`, `storage`, `path`, `path_hash`, `file`.`parent`, `name`, `mimetype`, `mimepart`, `size`, `mtime`, `storage_mtime`, `encrypted`, `etag`, `permissions`, `checksum`, `metadata_etag`, `creation_time`, `upload_time` FROM
`oc_filecache` `file` LEFT JOIN `oc_filecache_extended` `fe` ON `file`.`fileid` = `fe`.`fileid` WHERE (`storage` = 330) AND (((`path`  COLLATE utf8mb4_general_ci LIKE '__groupfolders/22/%') OR (`path_hash` = '…')) AND (`name`  COLLATE u
tf8mb4_general_ci LIKE '%…%')) ORDER BY `mtime` desc LIMIT 5;


+------+-------------+-------+--------+-----------------------------------------------------------------------------------------------------+----------+---------+---------------------+------+-------------+
| id   | select_type | table | type   | possible_keys                                                                                       | key      | key_len | ref                 | rows | Extra       |
+------+-------------+-------+--------+-----------------------------------------------------------------------------------------------------+----------+---------+---------------------+------+-------------+
|    1 | SIMPLE      | file  | index  | fs_storage_path_hash,fs_storage_mimetype,fs_storage_mimepart,fs_storage_size,fs_storage_path_prefix | fs_mtime | 8       | NULL                |   76 | Using where |
|    1 | SIMPLE      | fe    | eq_ref | PRIMARY                                                                                             | PRIMARY  | 8       | db_biss.file.fileid |    1 | Using where |
+------+-------------+-------+--------+-----------------------------------------------------------------------------------------------------+----------+---------+---------------------+------+-------------+

icewind1991 commented:
@nickvergessen make sure that you have #28476 and #28608 applied

skjnldsv commented:
/backport to stable22

skjnldsv commented:
/backport to stable21
